src/libpmemfile/preload.c

```diff
@@ -570,101 +388,109 @@ static long
+	return result;
 }
 
 static long
 hook_setxattr(long arg0, long arg1, long arg2, long arg3, long arg4,
 		int resolve_last)
 {
+	long result;
 	struct resolved_path where;
 
-	resolve_path(cwd_desc(), (const char *)arg0, &where,
+	struct vfd_reference at = pmemfile_vfd_at_ref(AT_FDCWD);
+
+	resolve_path(at, (const char *)arg0, &where,
 			resolve_last | NO_AT_PATH);
 
 	if (where.error_code != 0)
-		return where.error_code;
+		result = where.error_code;
+	else if (where.at_pool != NULL)
+		result = check_errno(-ENOTSUP, SYS_setxattr);
+	else
+		result = syscall_no_intercept(SYS_setxattr, where.path,
+						arg1, arg2, arg3, arg4);
 
-	if (!is_fda_null(&where.at.pmem_fda))
-		return check_errno(-ENOTSUP, SYS_setxattr);
+	pmemfile_vfd_unref(at);
 
-	return syscall_no_intercept(SYS_setxattr, where.path, arg1, arg2, arg3,
-			arg4);
+	return result;
 }
 
 static long
-hook_mkdirat(long fd, long path_arg, long mode)
+hook_mkdirat(int fd, long path_arg, long mode)
 {
 	long ret;
 	struct resolved_path where;
 
-	struct fd_desc at = fd_fetch(fd);
+	struct vfd_reference at = pmemfile_vfd_at_ref(fd);
 
 	resolve_path(at, (const char *)path_arg, &where, NO_RESOLVE_LAST_SLINK);
 
 	if (where.error_code != 0) {
 		ret = where.error_code;
-	} else if (is_fda_null(&where.at.pmem_fda)) {
+	} else if (where.at_pool == NULL) {
 		ret = syscall_no_intercept(SYS_mkdirat,
-		    where.at.kernel_fd, where.path, mode);
+		    where.at_kernel, where.path, mode);
 	} else {
-		long r = wrapper_pmemfile_mkdirat(where.at.pmem_fda.pool->pool,
-			where.at.pmem_fda.file, where.path, (mode_t)mode);
+		long r = wrapper_pmemfile_mkdirat(where.at_pool->pool,
+			where.at_dir, where.path, (mode_t)mode);
 
 		ret = check_errno(r, SYS_mkdirat);
 	}
 
-	fd_release(&at);
+	pmemfile_vfd_unref(at);
 
 	return ret;
 }
 
 static long
-openat_helper(long fd, struct resolved_path *where, long flags, long mode)
+openat_helper(struct resolved_path *where, long flags, long mode)
 {
-	PMEMfile *file = pmemfile_openat(where->at.pmem_fda.pool->pool,
-					where->at.pmem_fda.file,
+	PMEMfile *file = pmemfile_openat(where->at_pool->pool,
+					where->at_dir,
 					where->path,
 					((int)flags) & ~O_NONBLOCK,
 					(mode_t)mode);
 
-	log_write("pmemfile_openat(%p, %p, \"%s\", 0x%x, %u) = %p",
-					(void *)where->at.pmem_fda.pool->pool,
-					(void *)where->at.pmem_fda.file,
+	log_write("pmemfile_openat(%p, %p, \"%s\", 0x%x, %lu) = %p",
+					(void *)where->at_pool->pool,
+					(void *)where->at_dir,
 					where->path,
 					((int)flags) & ~O_NONBLOCK,
-					(mode_t)mode,
+					mode,
 					file);
 
-	if (file == NULL) {
-		(void) syscall_no_intercept(SYS_close, fd);
+	if (file == NULL)
 		return check_errno(-errno, SYS_openat);
-	}
-
-	util_mutex_lock(&fd_table_mutex);
 
-	__sync_add_and_fetch(&fd_table[fd].ref_count, 1);
-	fd_table[fd].pmemfile.pool = where->at.pmem_fda.pool;
-	fd_table[fd].pmemfile.file = file;
+	long fd = pmemfile_vfd_assign(where->at_pool, file, where->path);
```

Comment:
I think we should acquire file descriptor before calling pmemfile_open. With O_CREAT rolling back "open" would have to do more than just "close".

Reply:
That might make things a bit more complicated.
1) acquire an fd
2) call pmemfile_openat
3) associate the fd with the pmemfile_file

I would have to split pmemfile_vfd_assign to two steps.
Should that be done? Is it worth it?

Reply:
Also, probably we would have to hold the vfd_table_mutex during the whole thing.